rust: add new api for initialization

BennoLossin opened the pull request:
This is my newest version of the pinned-initialization API. You can find it as a stand-alone crate here: [pinned-init](https://crates.io/crates/pinned-init).

I plan to squash 276ae43ce0e96492860b71b75c7d7d9c1d728cba onto 50f21212555fda9de7362b2ef13da335b01d9fe8. I separated the commits for better earlier review.

I also need to provide a better commit message better explaining the API (linking to my blog *might* help, but it is a long read, so...).

I removed all `unsafe` initialization API from the `sync` module. If we still want to keep the old API and we want to deprecate it instead just let me know.

Closes #772

nbdd0121 commented on rust/kernel/sync/arc.rs:
```diff
@@ -97,6 +97,35 @@ impl<T> Arc<T> {
 
+    /// Tries to allocate a new [`Arc`] instance whose contents are not initialised yet.
+    pub fn try_new_uninit() -> Result<Arc<MaybeUninit<T>>> {
+        let layout = Layout::new::<ArcInner<MaybeUninit<T>>>();
```
Please make this code part of `UniqueArc::try_new_uninit`'s implementation. The `Arc::try_new_uninit` API makes no sense, given that unlike std's `Arc`, our `Arc` does not have `get_mut` so there logically an uninitialised `Arc` will be uninitialised forever.

wedsonaf commented on drivers/android/context.rs:
```diff
@@ -27,20 +31,15 @@ unsafe impl Sync for Context {}
-            manager: unsafe {
-                Mutex::new(Manager {
+        let ctx = UniqueArc::pin_init::<core::convert::Infallible>(pin_init!(Self {
+            manager: mutex!(
```
I think this should be called `new_mutex` or something else to that effect, just `mutex` is too generic. The same applies to the macros for other primitives.

nbdd0121 replied:
I would call it `mutex_new`.

BennoLossin replied:
I have named them `new_XXX` for the moment, I like this better, because they will be grouped in the imports together. We can bikeshed later.

nbdd0121 replied:
I suggest `XXX_new` because kernel C function always use that convention (and it sounds the same as `XXX::new`).

wedsonaf commented on drivers/android/context.rs:
```diff
@@ -27,20 +31,15 @@ unsafe impl Sync for Context {}
-            // SAFETY: Init is called below.
-            manager: unsafe {
-                Mutex::new(Manager {
+        let ctx = UniqueArc::pin_init::<core::convert::Infallible>(pin_init!(Self {
```
The reason this was a `Pin<UniqueArc>` originally was because we had to initialise the mutex in it. Since we don't need this anymore, couldn't this just be an `Arc`?

This also feels a bit too verbose (e.g., having to specify `core::convert::Infallible`). I haven't looked at the details yet, but it would be great if we could avoid this.

BennoLossin replied:
Currently I have not implemented `InPlaceInit` for `Arc`, since it does not fit the API (you can only call `pin_init` on it). But since we have full control over it, we could just add it as a method. Will do so 👍

BennoLossin replied:
And yeah sadly we have to specify the error type.

nbdd0121 replied:
This is just an unfortunate result of using `?` (which is why I chose to not have `?` in my `pin-init` crate but instead make error propagation automatic).

Basically since all error return paths use `?` which expands to a `From::from`, the error type never unifies with anything and stay as an inference variable, hence the explicit specification.

BennoLossin replied:
If we had the never type this would be a much more bearable `::<!>`

ojeda replied:
Does it work with the never type (even if it is not stable), or are you triggering some compiler bug etc. there?

BennoLossin replied:
It works with the never type

wedsonaf commented on drivers/android/node.rs:
```diff
@@ -63,37 +68,30 @@ pub(crate) struct NodeDeath {
-    /// The caller must call `NodeDeath::init` before using the notification object.
-    pub(crate) unsafe fn new(node: Arc<Node>, process: Arc<Process>, cookie: usize) -> Self {
-        Self {
+    #[allow(clippy::new_ret_no_self)]
```
Is there a way to avoid having to disable this for every function where we return `impl InInit<Self>`?

BennoLossin replied:
Not naming it `new`? I am not sure if we should even name them new, maybe `new_in_place` or something might be more expressive. AFAIK we could add `#![allow(clippy::new_ret_no_self)]` to `lib.rs`, but then it would not warn any longer in other places... Maybe we could get in touch with the clippy folks and ask for some workaround (I am not aware of any currently). There is precedent for this: https://github.com/rust-lang/rust-clippy/issues/3313

wedsonaf commented on drivers/android/process.rs:
```diff
@@ -712,14 +706,15 @@ impl Process {
+                    self.clone(),
+                    cookie,
+                ))
+                .map_err(|(i, _)| i)?,
```
What is the pair returned here?

BennoLossin replied:
It is `(E, UniqueArc<T>)` and `E` is `Infallible`.

the function called is `UniqueArc::<MaybeUninit<T>>::pin_init_now(self, impl PinInit<T, E>) -> Result<UniqueArc<T>, (E, Self)>` and if the initializer fails I wanted to give a way to still keep the allocated memory.

This is the only place this function is used, because I did not want to change the code. I thought that there might be a reason that the memory is first allocated and then some checks are run (lines 701-707). If we can refactor this code to use `UniqueArc::pin_init` then we could remove the `pin_init_now` function (which I would prefer).

wedsonaf commented on drivers/android/thread.rs:
```diff
@@ -241,23 +246,13 @@ impl Thread {
         let return_work = Arc::try_new(ThreadError::new(InnerThread::set_return_work))?;
         let reply_work = Arc::try_new(ThreadError::new(InnerThread::set_reply_work))?;
-        let mut thread = Pin::from(UniqueArc::try_new(Self {
+        let thread = UniqueArc::pin_init::<core::convert::Infallible>(pin_init!(Self {
```
This could also just be an `Arc` as the update performed below is through a lock.

wedsonaf commented on drivers/android/process.rs:
```diff
@@ -267,24 +272,13 @@ unsafe impl Sync for Process {}
 impl Process {
     fn new(ctx: Arc<Context>, cred: ARef<Credential>) -> Result<Arc<Self>> {
-        let mut process = Pin::from(UniqueArc::try_new(Self {
+        let process = UniqueArc::pin_init::<core::convert::Infallible>(pin_init!(Self {
```
Same as above, this can just be an `Arc`.

wedsonaf commented on drivers/android/transaction.rs:
```diff
@@ -88,26 +87,20 @@ impl Transaction {
             code: tr.code,
             flags: tr.flags,
-            data_size: tr.data_size as _,
+            data_size: tr.data_size as usize,
```
Is this needed?

(To clarify, I'm not going into the question of whether it is better/worse to be explicit about the type, we can discuss that elsewhere. I'm asking if the new macros made things such that the compiler can't infer the types anymore.)

BennoLossin replied:
Yes this is needed, otherwise we get a type inference error (have not investigated why).

wedsonaf reviewed (commented):
Just did a quick scan of the first few files and left a few comments.

This is a huge improvement over what we have now.

 (I'll allocate some time to go over the rest.)

ojeda commented on drivers/android/transaction.rs:
```diff
@@ -7,7 +7,10 @@ use kernel::{
     io_buffer::IoBufferWriter,
     linked_list::List,
     linked_list::{GetLinks, Links},
+    macros::pin_data,
+    pin_init,
```
Maybe these two should go into the prelude.

BennoLossin commented:
What is the reason that the CI checks were cancelled?

wedsonaf commented:
> What is the reason that the CI checks were cancelled?

Not sure. Let me ask them to rerun.

ojeda commented on rust/kernel/init/__private.rs:
```diff
@@ -0,0 +1,123 @@
+//! Workaround for specialization
+use super::*;
```
```suggestion
// SPDX-License-Identifier: GPL-2.0

//! Workaround for specialization.

use super::*;
```

Similarly in `init.rs` and `pin_data.rs`.

ojeda commented on rust/kernel/init.rs:
```suggestion
/// Initialize a pinned type on the stack.
///
/// # Examples
///
/// ```rust
```

ojeda commented on rust/kernel/init.rs:
```suggestion
/// # Safety
///
/// The bit pattern consisting of only zeroes must be a valid bit pattern for the type.
```

(and similar elsehwere)

ojeda commented on rust/kernel/init.rs:
```suggestion
// We are allowed to zero padding bytes.
```

(and similarly elsewhere)

ojeda commented on rust/kernel/lib.rs:
```diff
@@ -28,12 +28,20 @@
+// TODO resolve discussion
+#![feature(new_uninit)]
+#![feature(type_alias_impl_trait)]
+// these are already stabilized in newer versions
```
Thanks! I will track it in the list -- you can remove this comment.

ojeda commented on rust/kernel/init.rs:
```suggestion
/// ```
///
/// Users of `Foo` can now create it like this:
///
/// ```rust
```

(and similarly elsewhere)

ojeda commented on rust/kernel/init.rs:
Should we use the convenience attribute in the examples? (i.e. users will typically use that one instead, right?)

BennoLossin replied:
yes, forgot to add this, since I do not have the convenience attr in my crate

ojeda commented on rust/macros/lib.rs:
```diff
@@ -189,3 +190,9 @@ pub fn vtable(attr: TokenStream, ts: TokenStream) -> TokenStream {
     concat_idents::concat_idents(ts)
 }
+
+/// `pin_data!` convenience attribute
```
```suggestion
/// `pin_data!` convenience attribute.
///
/// It allows to avoid having to wrap struct definitions in `pin_data! { ... }`.
///
/// # Examples
///
/// ```rust
/// #[pin_data]
/// struct A;
/// ```
```

ojeda replied:
Given `pin_data!` will be hidden, I would avoid mentioning it at all.

nbdd0121 commented:
My branch that uses `pin-init` crate for surface syntax comparison https://github.com/nbdd0121/linux/commit/a1178f850cdc0e68db81e15164ede6c6c006d73e.

ojeda commented on rust/kernel/init.rs:
I would probably mention `#[pin_data]` here if that is what users will normally use.

Actually, is there some case where users would need `pin_data!` directly? If not, should we make this `#[doc(hidden)]`?

BennoLossin replied:
not that I can think of... so we should hide it

nbdd0121 commented:
@y86-dev it seems that the init syntax has been further changed within the past couple of days? The surface syntax is very similar to `pin-init` now. Nice :)

One difference is that the thing within `pin_init!` is not quite a struct expression, when the type has generic involved. E.g. If we have `Foo<T>`, then currently you have to `pin_init!(Foo<T> { xxx })` instead of `pin_init!(Foo { xxx })`.

BennoLossin commented:
Not sure which part of the syntax you are referring to... If you look at https://github.com/y86-dev/pinned-init/blame/main/examples/mutex.rs#L68 that has not changed for 22 days.

> One difference is

Yeah, I need that for the struct initializer (otherwise again some inference error).

ojeda reviewed (commented):
Some nits from a quick look for the next time you force-push.

By the way, the commit splitting looks fine to me! Why do you want to squash the two commits?

As for the commit message and the:

```
TODO: describe the purpose and functionality of the API better.
```

I am guessing you may be dreading explaining everything there :) If that is the case, do not worry too much about explaining everything in the commit message. In particular, you don't need to repeat the docs/API since it will be in the source code. Instead, try to focus on the "why do we want/need this change".

BennoLossin commented:
> Some nits from a quick look for the next time you force-push.

I plan to do normal commits to have a better review and will rebase at the end

nbdd0121 commented:
> Not sure which part of the syntax you are referring to... If you look at [y86-dev/pinned-init@`main`/examples/mutex.rs#L68 (blame)](https://github.com/y86-dev/pinned-init/blame/main/examples/mutex.rs#L68) that has not changed for 22 days.

Oh, sorry! I was under the impression that the syntax on the README was the latest syntax when I checked last time.

> Yeah, I need that for the struct initializer (otherwise again some inference error).

That's quite unfortunate. I am a big fan of your approach (todo struct, and the raw-pointer-based API), but I think the current state as it is requires too many explicit annotations.

BennoLossin commented:
> Oh, sorry! I was under the impression that the syntax on the README was the latest syntax when I checked last time.

Did you mean the syntax with `<-`? That was just experimental to see how it could look like as a language feature (I also let the readme go stale...).

nbdd0121 commented:
One idea about solving the error type inference problem: given that in kernel we usually only use two error types, `!` and `kernel::Error`, maybe we can have two macros, one for each error type? (the suggestion won't help the library for general audience, but should be good enough for us)

We can call the macros and functions like this:
* `init!` for infallible pin-init. (we can drop `pin` from the name too because in kernel we probably wouldn't need to initialize a struct unpinned; if people need to initialize `Unpin` struct then they can pin-init it first and then undo the pinning)
* `try_init!` for pin-init with `kernel::Error` return type.
* `{Box, UniqueArc}::{init_with, try_init_with}` that initialize a `Ptr<MaybeUninit<T>>` with `!` and `Error` respectively.
* `{Box, UniqueArc, Arc}::try_new_with` that takes both `PinInit<T, !>` and `PinInit<T, Error>` (because allocation fail will return an error anyway).

Of course this doesn't help with the inference problem at the struct and field...

bjorn3 commented on rust/kernel/lib.rs:
```diff
@@ -28,12 +28,20 @@
 #![feature(receiver_trait)]
 #![feature(unsize)]
+// TODO resolve discussion
+#![feature(new_uninit)]
```
This one should be fine.

ojeda replied:
I will add it to the list.

bjorn3 commented on rust/kernel/lib.rs:
```diff
@@ -28,12 +28,20 @@
 #![feature(unsize)]
+// TODO resolve discussion
+#![feature(new_uninit)]
+#![feature(type_alias_impl_trait)]
```
I'm less sure about this one. As I understand it there are still outstanding issues (ICE, unsoundness, unexpected behavior) with it that are being fixed: https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3AF-type_alias_impl_trait

BennoLossin replied:
yeah, sadly I cannot really work around this one... I need it for the `LockFactory` trait. Maybe it can be done in a different way, but I cannot name a type that implements `PinInit`/`Init`, because it contains a closure...

bjorn3 replied:
What about writing a custom type implementing `PinInit` rather than `PinInitClosure` with a closure?

BennoLossin replied:
How would that work? I somehow need to have different behavior for each call site of `pin_init!`.

bjorn3 replied:
You are talking about `Mutex::new`, right? You could do the following I think:

```rust
struct RawMutexPinInit {
    name: &'static CStr,
    key: &'static LockClassKey,
}

unsafe impl<T, E> PinInit<Opaque<bindings::mutex>, E> for RawMutexPinInit
{
    unsafe fn __pinned_init(self, slot: *mut Opaque<bindings::mutex>) -> Result<(), E> {
        bindings::__mutex_init(Opaque::raw_get(slot), self.name.as_char_ptr(), self.key.get());
        Ok(())
    }
}

pub const fn new(t: T, name: &'static CStr, key: &'static LockClassKey) -> impl PinInit<Self> {
    pin_init!(Self {
        mutex: RawMutexPinInit { name, key },
        data: UnsafeCell::new(t),
        _pin: PhantomPinned,
    })
}
```

---

After writing this comment I realized you meant the `pin_init!` call, not `init_mutex()` used inside `pin_init!`. In any case you could do the same transformation for the `pin_init!` I think after macro expanding it, or introduce a `pin_init!` alternative which allows you to explictly declare the `PinInit` implementor name combined with all used upvars. Something like

```rust
pin_init_concrete!(
    // Let pin_init_concrete define a MutexPinInit type which implements PinInit
    MutexPinInit<T> { t: T, name: &'static CStr, key: &'static LockClassKey };
    Self {
        mutex: RawMutexPinInit { name: self.name, key: self.key },
        data: UnsafeCell::new(self.t),
        _pin: PhantomPinned,
    }
)
```

It is harder to write though.

BennoLossin replied:
How would this work for `SeqLock` though?
```rust
    pub const fn new(
        data: L::Inner,
        name: &'static CStr,
        key1: &'static LockClassKey,
        key2: &'static LockClassKey,
    ) -> impl PinInit<Self, L::Error>
    where
        L: LockFactory<LockedType<L::Inner> = L>,
        L::Inner: Sized,
        <L as LockFactory>::Error: From<core::convert::Infallible>,
    {
        fn init_count(
            name: &'static CStr,
            key2: &'static LockClassKey,
        ) -> impl PinInit<Opaque<bindings::seqcount>> {
            let init = move |place: *mut Opaque<bindings::seqcount>| {
                unsafe {
                    bindings::__seqcount_init(
                        Opaque::raw_get(place),
                        name.as_char_ptr(),
                        key2.get(),
                    )
                };
                Ok(())
            };
            unsafe { PinInitClosure::from_closure(init) }
        }
        pin_init!(Self {
            _p: PhantomPinned,
            count: init_count(name, key2),
            write_lock: L::new_lock(data, name, key1),
        })
    }
```

As it needs to be generic, the macro you are proposing seems *very* complicated...

bjorn3 replied:
If we had syn it wouldn't be so hard to parse :P On a more serious note, I agree, it is not ideal. I'm just worried people will hit the bugs in type_alias_impl_trait.

BennoLossin replied:
For sure! I have not encountered a single ICE, so we might only use what is currently implemented correctly...

BennoLossin replied:
But I think we could use your approach for the types that implement `LockFactory` and then we would not need this feature. I will try and implement it after I fixed some of the other things.

bjorn3 commented on rust/kernel/lib.rs:
```diff
@@ -28,12 +28,20 @@
 compile_error!("Missing kernel configuration for conditional compilation");
 
+// allow macros to refer to `::kernel`.
+extern crate self as kernel;
```
I don't see any references to it in any macro.

BennoLossin replied:
my `pin_data` macro uses it, I should clarify that this only applies to proc-macros.

bjorn3 commented on rust/kernel/init.rs:
Maybe add a comment that this is the drop guard. At first I got confused and thought it was the actual initialization value.

BennoLossin commented:
> One idea about solving the error type inference problem: given that in kernel we usually only use two error types, `!` and `kernel::Error`, maybe we can have two macros, one for each error type? (the suggestion won't help the library for general audience, but should be good enough for us)
> 
> We can call the macros and functions like this:
> 
>     * `init!` for infallible pin-init. (we can drop `pin` from the name too because in kernel we probably wouldn't need to initialize a struct unpinned; if people need to initialize `Unpin` struct then they can pin-init it first and then undo the pinning)
> 
>     * `try_init!` for pin-init with `kernel::Error` return type.
> 
>     * `{Box, UniqueArc}::{init_with, try_init_with}` that initialize a `Ptr<MaybeUninit<T>>` with `!` and `Error` respectively.
> 
>     * `{Box, UniqueArc, Arc}::try_new_with` that takes both `PinInit<T, !>` and `PinInit<T, Error>` (because allocation fail will return an error anyway).
> 
> 
> Of course this doesn't help with the inference problem at the struct and field...

I am not sure... Users might want to have custom errors that are ZSTs or have more options than `error::Error` has.

There *could* be a situation where you want to initialize something `!Unpin` in-place and then later move it (although I cannot think of anything and this also came up in the internals thread going over my blog.). Not sure if we want to shut this door early and have a lot of refactoring later, or let it stay open and... also have a lot of refactoring...

bjorn3 commented:
Another option to deal with error type inference would be to allow explicitly specifying the error type in the macro if it couldn't be inferred.

bjorn3 commented on rust/kernel/init.rs:
```suggestion
/// #[pin_project(PinnedDrop)]
```
right?

BennoLossin replied:
Yeah... Nice find! I think I will make it an error to implement `PinnedDrop` on a struct with `#[pin_project]` instead of `#[pin_project(PinnedDrop)]`.

BennoLossin commented:
Summary of the last few commits (up to the last force-push):

- back to declarative macro + convenience attribute macro.
- the attr macro actually does an important parsing step: separating `impl_generics` and `ty_generics` and isolate the `where` clause
- better docs
- prevention of `Drop` and `Unpin` impls
- `PinnedDrop`
- other small changes (the explicit `PinInitClosure` type is gone)

BennoLossin commented:
rebased commits again that had been accumulating. Also fixed some unsoundness with `DropGuard` (it was possible to forget it and then error anyway). Also rebased with the newest changes from `rust`.

bjorn3 commented on drivers/android/process.rs:
```diff
@@ -712,14 +703,15 @@ impl Process {
+                    self.clone(),
+                    cookie,
+                ))
+                .map_err(|(i, _)| i)?,
```
```suggestion
                .map_err(|(err, _)| err)?,
```
or
```suggestion
                .unwrap_or_else(|(err, _)| match err {}),
```

bjorn3 commented on rust/kernel/sync.rs:
```diff
@@ -131,35 +110,70 @@ macro_rules! init_static_sync {
+    inner: MaybeUninit<UnsafeCell<T>>,
+}
+
+unsafe impl<T: Sync> Sync for StaticInit<T> {}
+unsafe impl<T: Send> Send for StaticInit<T> {}
```
Missing safety comments.

bjorn3 commented on rust/macros/pinned_drop.rs:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+    let idx = pinned_drop_idx.unwrap();
+    //inserting `::kernel::init::` in reverse order
```
Maybe [`Vec::splice`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.splice) with `idx..idx` as range would work?

bjorn3 replied:
https://github.com/Rust-for-Linux/linux/pull/909/commits/d1a1330fd11ea088fb49e286002da434169b1c55 didn't start using splice unlike what the commit title suggests.

BennoLossin replied:
oh wrong directory in git...

bjorn3 commented on rust/kernel/sync.rs:
```diff
@@ -154,11 +158,13 @@ impl<T> StaticInit<T> {
     ///
     /// The caller calls this function exactly once and before any other function (even implicitly
-    /// derefing) of `self` is called.
+    /// derefing) of `self` is called. `self` stays pinned indefinetly.
```
Maybe it should take `Pin<&Self>` or `&'static Self`? `Pin::static_ref` can be used to turn `&'static T` into `Pin<&'static T>` without unsafe code, so it doesn't really matter which of the two is used. Also typo here and elsewhere:
```suggestion
    /// derefing) of `self` is called. `self` stays pinned indefinitely.
```

BennoLossin commented on drivers/android/process.rs:
```diff
@@ -712,14 +703,15 @@ impl Process {
-            pinned.as_mut().init();
-            Arc::<NodeDeath>::from(pinned)
-        };
+        let death = Arc::<NodeDeath>::from(
+            death
+                .pin_init_now::<core::convert::Infallible>(NodeDeath::new(
+                    info.node_ref.node.clone(),
+                    self.clone(),
+                    cookie,
+                ))
+                .unwrap_or_else(|(err, _)| match err {}),
+        );
```
@wedsonaf can we refactor this code? So that we can merge lines 693-696
```rust
        let death = UniqueArc::try_new_uninit().map_err(|err| {
            thread.push_return_work(BR_ERROR);
            err
        })?;
```
with this?

Then I could remove `pin_init_now` and just use `pin_init` here.

wedsonaf replied:
The problem with such a merge is that we'd have the memory allocation path (which may block for a while if the OOM killer is triggered) while holding a mutex. This is ok wrt safety, but is different from what the C version does.

It is unclear to me if the C version did this as some targeted optimisation after some profiling or if it's some potentially premature optimisation. I mimicked to C version here to reduce the risk of regressions.

Anyway, since we are unclear about the future of this codebase, I think it is ok to refactor for now. It would be nice if we had a long-term plan for supporting this scenario though.

BennoLossin replied:
Oh I totally overlooked that `lock()` call. I think we should keep it this way.

nbdd0121 commented on rust/kernel/sync/arc.rs:
```diff
@@ -464,9 +464,33 @@ impl<T> UniqueArc<T> {
+        // assert that there are only two fields: refcount and data (done in a closure to avoid
+        // overflowing the stack in debug mode with a big `T`)
+        #[allow(unreachable_code, clippy::diverging_sub_expression)]
+        let _check = || {
```
I don't feel that this check is necessary given that the definition of `ArcInner` is in the same file.

BennoLossin replied:
What do you think of the implementation now?

nbdd0121 commented on rust/kernel/sync/arc.rs:
```diff
@@ -464,9 +464,33 @@ impl<T> UniqueArc<T> {
+        let inner = NonNull::new(unsafe { alloc(layout) })
+            .ok_or(ENOMEM)?
+            .cast::<ArcInner<MaybeUninit<T>>>();
+        // TODO do this using `pinned-init`
```
Nit: I don't think we need this TODO.

nbdd0121 commented on rust/kernel/types.rs:
```diff
@@ -316,6 +316,18 @@ impl<T> Opaque<T> {
+    ///
+    /// The pointer supplied is valid.
+    pub unsafe fn raw_get(this: *const Self) -> *mut T {
+        UnsafeCell::raw_get(addr_of!((*this).0).cast::<UnsafeCell<T>>())
```
Maybe
```rust
UnsafeCell::raw_get(this.cast::<UnsafeCell<T>>())
```
? `Opaque` and `MaybeUninit` are both `#[repr(transparent)]` so we can just cast the pointer directly.

BennoLossin replied:
Seems reasonable, it was a bit unclear to me if we guarantee that `Opaque` will always be `transparent` (I guess changing it would also require changing this method, but wanted to be careful). To be clear: I do not think that there is a reason for making `Opaque` not transparent, so we could just clarify that in the documentation.

Darksonn replied:
Normally, the `UnsafeCell` would go _outside_ the `MaybeUninit`.

nbdd0121 commented on rust/kernel/init/pinned_drop.rs:
```diff
@@ -0,0 +1,29 @@
+                $($inner)*
+            }
+
+            fn __ensure_no_unsafe_op_in_drop($self: $st) {
```
I don't think this is necessary, we enable `unsafe_op_in_unsafe_fn` globally. If you want to be extra careful you can generate a `#[forbid(unsafe_op_in_unsafe_fn)]` on the drop function.

bjorn3 replied:
You can still silence forbid using --cap-lints allow.

BennoLossin replied:
I think we should be extra careful here.

nbdd0121 commented on drivers/android/node.rs:
```diff
@@ -63,37 +66,30 @@ pub(crate) struct NodeDeath {
-    /// The caller must call `NodeDeath::init` before using the notification object.
-    pub(crate) unsafe fn new(node: Arc<Node>, process: Arc<Process>, cookie: usize) -> Self {
-        Self {
+    #[allow(clippy::new_ret_no_self)]
```
We probably want to disable this globally until https://github.com/rust-lang/rust-clippy/issues/7344 is fixed.

nbdd0121 replied:
Filed https://github.com/rust-lang/rust-clippy/pull/9733 to fix the false positive.

BennoLossin replied:
Should I just add `#![allow(clippy::new_ret_no_self)]` to `lib.rs`?

nbdd0121 commented on drivers/android/process.rs:
```diff
@@ -267,25 +270,13 @@ unsafe impl Sync for Process {}
-        kernel::mutex_init!(pinned, "Process::node_refs");
-
-        Ok(process.into())
+            task: ARef::from(Task::current().group_leader()),
```
Is this change caused by an inference issue?

BennoLossin replied:
yes

nbdd0121 commented on rust/kernel/sync/arc.rs:
```diff
@@ -475,6 +512,44 @@ impl<T> UniqueArc<MaybeUninit<T>> {
+    pub fn pin_init_now<E>(
+        mut self,
+        init: impl PinInit<T, E>,
+    ) -> core::result::Result<Pin<UniqueArc<T>>, (E, UniqueArc<MaybeUninit<T>>)> {
```
Would clippy still warn if this was `Result<Pin<UniqueArc<T>>, (E, Self)>`?

BennoLossin replied:
Nope 👍

nbdd0121 commented on rust/kernel/sync/arc.rs:
```diff
@@ -475,6 +512,44 @@ impl<T> UniqueArc<MaybeUninit<T>> {
+    }
+
+    /// Initialize the value after creating the `UniqueArc`.
+    pub fn init_now<E>(
```
Is this really necessary? I suppose we can just have a pinned variant, and if someone needs an unpinned version then they can take it out from `Pin` themself.

Note that our `Arc` is different from Rust std `Arc`, it is pinned by default. Our `UniqueArc` isn't pinned by default, but we probably could. @wedsonaf do you think it make sense to have `UniqueArc` pinned and only provide a `DerefMut` if `T: Unpin`? We can add an `as_mut` function that gives `Pin<&mut Self>` for `T: !Unpin`.

BennoLossin replied:
I added it for completeness. We could surely just remove it. However it might occur that someone has some data that is `!Unpin` and they want to still be able to mutate it before pinning... (no idea how likely that is though, probably very unlikely)

nbdd0121 commented on rust/kernel/sync/condvar.rs:
```diff
@@ -44,15 +58,27 @@ unsafe impl Sync for CondVar {}
-            _pin: PhantomPinned,
+    #[allow(clippy::new_ret_no_self)]
+    pub const fn new(name: &'static CStr, key: &'static LockClassKey) -> impl PinInit<Self> {
+        fn init_wait_list(
```
What's the rationale behind having a new function instead of just having
```rust
wait_list: unsafe { init::pin_init_from_closure(|| { ... }) }
```
down there?

BennoLossin replied:
here would be the closure version:
```rust
        pin_init!(Self {
            wait_list: unsafe {
                init::pin_init_from_closure::<_, core::convert::Infallible>(move |slot| {
                    bindings::__init_waitqueue_head(
                        Opaque::raw_get(slot),
                        name.as_char_ptr(),
                        key.get(),
                    );
                    Ok(())
                })
            },
            _pin: PhantomPinned,
        })
```
(yes we need to specify the error type 🙄)

Not sure if that is more readable... I could probably create a more general function that would handle this, so along the lines of:
```rust
unsafe fn ffi_init1<T, P1>(fun: unsafe fn(*mut T, P1), param1: P1) -> impl PinInit<Opaque<T>>;
unsafe fn ffi_init2<T, P1, P2>(fun: unsafe fn(*mut T, P1), param1: P1, param2: P2) -> impl PinInit<Opaque<T>>;
unsafe fn ffi_init3<T, P1, P2, P2>(fun: unsafe fn(*mut T, P1), param1: P1, param2: P2, param3: P3) -> impl PinInit<Opaque<T>>;
```

nbdd0121 commented on rust/macros/pin_project.rs:
```diff
@@ -0,0 +1,111 @@
+                        TokenTree::Punct(p) if p.as_char() == ',' => at_start = true,
+                        TokenTree::Punct(p) if p.as_char() == '\'' && at_start => {
+                            ty_generics.push(tt.clone());
+                            ty_generics.push(TokenTree::Punct(Punct::new(',', Spacing::Alone)));
```
Wouldn't this push `',` into `ty_generics` instead `'lifetime,`?

I think the comma should be pushed only when we see a comma.

BennoLossin replied:
oh yes that is correct

nbdd0121 commented on rust/macros/pin_project.rs:
```diff
@@ -0,0 +1,111 @@
+    }
+    rest.extend(toks);
+    let last = rest.pop();
+    TokenStream::from_iter(vec![
```
I think this could be more readable if written as:
```rust
let mut ret = TokenStream::new();
ret.extend("::kernel::pin_project!".parse());
...
```

nbdd0121 commented on rust/macros/pinned_drop.rs:
```diff
@@ -0,0 +1,82 @@
+    let mut pinned_drop_idx = None;
+    for (i, tt) in toks.iter().enumerate() {
+        match tt {
+            TokenTree::Punct(p) if p.as_char() == '<' => {
```
Perhaps the angle bracket matching code can be extracted to a function and shared between both macros.

BennoLossin replied:
This seems a bit difficult, because in the code for `pin_project` I am going through the generics list and I am searching for the identifiers. So that code seems a bit too specific to extract that and keep the functionality...

I can still create the function for use in other places though...

nbdd0121 commented on rust/macros/pinned_drop.rs:
```diff
@@ -0,0 +1,82 @@
+    // fully qualify the `PinnedDrop`.
+    toks.splice(
+        idx..idx,
+        vec![
```
`"::kernel::init::".parse::<TokenStream>()`

nbdd0121 commented:
I had a brief look and it does indeed seem like the inference will be very limited with current design.

However I think maybe we can have slightly different syntax for direct initialization vs in-place initialization, e.g.
```rust
pin_init!(Self {
    foo: 1, // this is direct initialization
    mutex <- new_mutex!(...), // this is in-place initialization
})
```
this way type inference can work properly for direct initialization, and we also have a clear indication of which one is pinned initialized and which one is not (could be a good thing potentially?)

BennoLossin commented:
This was one of the intermediate states I had, but I changed it, because at kangrejos there were some worries about non-standard syntax. I think that it makes it easier to understand (and as you also said makes inference better) how things are initialized.

BennoLossin commented:
closed since <https://lore.kernel.org/rust-for-linux/20230408122429.1103522-1-y86-dev@protonmail.com/> was merged
